pkg/devfile/parser/parse_test.go

```diff
@@ -4155,6 +4162,280 @@ func Test_parseFromURI(t *testing.T) {
+		name            string
+		curDevfileCtx   *devfileCtx.DevfileCtx
+		gitUrl          *git.Url
+		importReference v1.ImportReference
+		wantDevFile     DevfileObj
+		wantError       *string
+	}{
+		{
+			name:          "private main devfile URL",
+			curDevfileCtx: &curDevfileContextWithValidToken,
+			gitUrl:        privateGitUrl,
+			importReference: v1.ImportReference{
+				ImportReferenceUnion: v1.ImportReferenceUnion{
+					Uri: server.URL,
+				},
+			},
+			wantDevFile: minimalDevfile,
+		},
+		{
+			name:          "private main devfile Bitbucket URL",
+			curDevfileCtx: &curDevfileContextWithValidToken,
+			gitUrl:        privateBitbucketGitUrl,
+			importReference: v1.ImportReference{
+				ImportReferenceUnion: v1.ImportReferenceUnion{
+					Uri: server.URL,
+				},
+			},
+			wantDevFile: minimalDevfile,
+		},
+		{
+			name:          "private main devfile with a private parent reference",
+			curDevfileCtx: &curDevfileContextWithValidToken,
+			gitUrl:        privateGitUrl,
+			importReference: v1.ImportReference{
+				ImportReferenceUnion: v1.ImportReferenceUnion{
+					Uri: parent.URL,
+				},
+			},
+			wantDevFile: parentDevfile,
+		},
+		{
+			name:          "private main devfile with a private parent with a nested private parent reference",
+			curDevfileCtx: &curDevfileContextWithValidToken,
+			gitUrl:        privateGitUrl,
+			importReference: v1.ImportReference{
+				ImportReferenceUnion: v1.ImportReferenceUnion{
+					Uri: nestedParent.URL,
+				},
+			},
+			wantDevFile: nestParentDevfile,
+		},
+		{
+			name:          "private main devfile without a valid token",
+			curDevfileCtx: &curDevfileContextWithInvalidToken,
+			gitUrl:        privateGitUrl,
+			importReference: v1.ImportReference{
+				ImportReferenceUnion: v1.ImportReferenceUnion{
+					Uri: server.URL,
+				},
+			},
+			wantError: &invalidTokenError,
+		},
+		{
+			name:          "public main devfile without a token",
+			curDevfileCtx: &curDevfileContextWithoutToken,
+			gitUrl:        publicGitUrl,
+			importReference: v1.ImportReference{
+				ImportReferenceUnion: v1.ImportReferenceUnion{
+					Uri: server.URL,
+				},
+			},
+			wantDevFile: minimalDevfile,
+		},
+		{
+			name:          "private parent devfile with invalid revision",
+			curDevfileCtx: &curDevfileContextWithoutToken,
+			gitUrl:        privateGitUrlInvalidRevision,
+			importReference: v1.ImportReference{
+				ImportReferenceUnion: v1.ImportReferenceUnion{
+					Uri: parent.URL,
+				},
+			},
+			wantError: &invalidGitSwitchError,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			downloadGitRepoResources = mockDownloadGitRepoResources(tt.gitUrl)
+			got, err := parseFromURI(tt.importReference, *tt.curDevfileCtx, &resolutionContextTree{}, resolverTools{})
+			if (err != nil) != (tt.wantError != nil) {
+				t.Errorf("Unexpected error: %v, wantErr %v", err, tt.wantError)
+			} else if err == nil && !reflect.DeepEqual(got.Data, tt.wantDevFile.Data) {
```

Comment:
It's not just the parsed devfile we want to validate, we also want to validate your new changes too, specifically the output of the clone function.  

In the mockExecute function, you return content "test" if a repo is successfully cloned.  If it has a revision that you can switch to, the expected content then becomes "git switched to revision". You should test out these combinations: 
 1.  If the git URL is valid and has no revisions, then validate the output is "test"
 2. If the git URL has an invalid revision then make sure destination directory is empty since the entire operation should fail and you will need to clean up.  I think there should be a parse error here too if the parent devfile can't be determined.  Just to point out for mocking purposes,  we are forcing `tt.gitUrl` to be the parent URL since the real URL we set up as the import reference is not a valid git provider (it's just a localhost server)
 3. If the git URL has a valid revision, then the content should be "git switched to revision"